apps/explorer/lib/explorer/third_party_integrations/universal_proxy.ex

```diff
@@ -0,0 +1,322 @@
+  defp parse_method(nil), do: nil
+
+  # sobelow_skip ["DOS.StringToAtom"]
+  defp parse_method(platform_config) do
+    raw_method = String.to_atom(String.downcase(platform_config["endpoints"]["base"]["method"]))
+
+    if raw_method in @allowed_methods do
+      raw_method
+    else
+      nil
+    end
+  end
+
+  @spec parse_endpoint_api_key(api_request(), api_key() | nil, String.t() | nil) :: api_request()
+  defp parse_endpoint_api_key(%{url: _url, headers: _headers} = map, nil, _platform_id), do: map
+
+  defp parse_endpoint_api_key(
+         %{url: url, headers: headers} = map,
+         %{"location" => location, "param_name" => param_name} = endpoint_api_key,
+         platform_id
+       ) do
+    api_key_env_name = "UNIVERSAL_PROXY_" <> String.upcase(platform_id) <> "_API_KEY"
+    endpoint_api_key_value = System.get_env(api_key_env_name)
+
+    case location do
+      "header" ->
+        if endpoint_api_key["prefix"] do
+          Map.put(map, :headers, [
+            {param_name, "#{endpoint_api_key["prefix"]} #{endpoint_api_key_value}"} | headers
+          ])
+        else
+          Map.put(map, :headers, [{param_name, endpoint_api_key_value} | headers])
+        end
+
+      "query" ->
+        Map.put(map, :url, url <> "?#{param_name}=#{endpoint_api_key_value}")
+
+      _ ->
+        map
+    end
+  end
+
+  defp parse_endpoint_api_key(map, _endpoint_api_key, _platform_id), do: map
+
+  @spec parse_endpoint_params(api_request(), [endpoint_param()], map()) :: api_request()
+  defp parse_endpoint_params(api_request_map, endpoint_params, proxy_params) do
+    endpoint_params
+    |> Enum.reduce(api_request_map, fn param, map ->
+      case param do
+        %{"location" => _location, "name" => _param_name, "value" => value} ->
+          parse_param_location(map, param, value)
+
+        %{"location" => _location, "type" => "chain_id_dependent", "mapping" => mapping} ->
+          parse_param_location(map, param, mapping[proxy_params["chain_id"]])
+
+        %{"location" => _location, "type" => param_type} ->
+          parse_param_location(map, param, proxy_params[param_type])
+
+        _ ->
+          map
+      end
+    end)
+  end
+
+  @spec parse_param_location(api_request(), endpoint_param(), String.t()) :: api_request()
+  defp parse_param_location(
+         %{body: body, url: url, headers: headers} = api_request_map,
+         %{"location" => location} = params,
+         value
+       )
+       when not is_nil(url) and not is_nil(value) do
+    case location do
+      "path" ->
+        if value do
+          Map.put(api_request_map, :url, String.replace(url, ":#{param_name(params)}", value))
+        else
+          api_request_map
+        end
+
+      "query" ->
+        query_param_name = param_name(params)
+        delimiter = if String.contains?(url, "?"), do: "&", else: "?"
+        Map.put(api_request_map, :url, url <> "#{delimiter}#{query_param_name}=#{value}")
```

Comment:
I believe the current approach is enough for this case. However, as an idea, we could leverage the URI module here. Tests are passing locally, but feel free to skip this if you prefer the existing solution.

```suggestion

         updated_uri =
          url
          |> URI.parse()
          |> URI.append_query("#{query_param_name}=#{value}")
          |> URI.to_string()
          |> URI.encode()
  
        Map.put(api_request_map, :url, updated_uri)
```